pkg/ccl/changefeedccl/sink_external_connection.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/cloud/externalconn"
+	"github.com/cockroachdb/cockroach/pkg/cloud/externalconn/connectionpb"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/errors"
+)
+
+func makeExternalConnectionSink(
+	ctx context.Context,
+	u sinkURL,
+	user username.SQLUsername,
+	db *kv.DB,
+	ie sqlutil.InternalExecutor,
+	sinkCtx SinkContext,
+) (Sink, error) {
+	if u.Host == "" {
+		return nil, errors.Newf("host component of an external URI must refer to an "+
+			"existing External Connection object: %s", u.String())
+	}
+
+	externalConnectionName := u.Host
+
+	// TODO(adityamaru): Use the `user` in `cfg` to perform privilege checks on
+	// the external connection object we are about to retrieve.
+
+	// Retrieve the external connection object from the system table.
+	var ec *externalconn.ExternalConnection
+	if err := db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		var err error
+		ec, err = externalconn.LoadExternalConnection(ctx, externalConnectionName,
+			connectionpb.TypeStorage, ie, user, txn)
+		return err
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to load external connection object")
+	}
+
+	// Construct a Sink handle for the underlying resource represented by the
+	// external connection object.
+	details := ec.ConnectionDetails()
+	connDetails, err := externalconn.MakeConnectionDetails(ctx, *details)
+	if err != nil {
+		return nil, err
+	}
+	connection, err := connDetails.Dial(ctx, sinkCtx, "" /* subdir */)
```

Comment:
Please don't dial here.  We previously removed this coupling for all sinks (ie. create just creates the sink, and Dial -- dials).  Definitely don't want to be making external
connections under txn...

Reply:
Actually, this external *kafka* connection is a bit confusing.
Why not just have external connection?

If the above call to Dial is okay -- i.e. we're not dialing kafka cluster, we're just doing a lookup on external conn table, then ... once you have connDetails, why can't we just unwrap it, yank URI out of there and recursively call  getSink() just with different sink ULI?

On the other hand, if we don't want to have this lookup to happen at all, then
I still don't understand why we need kafka external connection.
We could just have something like:

```
type externalConn struct {
  name string
  ....
}

// externalConn implements external connection methods: Dial, etc...

type externalConnSink struct {
  conn externalConn
  getSink func(uri string)
  wrapped Sink
}

// This is *Sink* dial
func (s *externalConnSink) Dial() error {
   // Dial external conn
   s.conn.Dial() ...
   // Determine underlying uri; getSink is a callback to getSink method
   s.wraped = s.getSink(realURI)
  ...
}

// All other sink methods delegate to s.wrapped.X
```

Reply:
> Definitely don't want to be making external connections under txn

ConnectionDetails.Dial differs from the `DIal` in `externalResource` in that it doesn't make any connection. It only hands you back a `Sink` interface which is what we already do in `getSink`? This Dial does basically what you described:
```
once you have connDetails, why can't we just unwrap it, yank URI out of there, and recursively call getSink() just with different sink ULI?
```

The interface method goes to the correct constructor based on the externalConnDetails provider which is what we register in `sink_kafka_connection.go`. The particular constructor knows how to unmarshal the connDetails and what to pick from it. It then calls the specific sinks constructor method as is called in `sink.go`. ConnctionDetails.Dial is meant to be the method that one uses to create a handle on the underlying resource. Doing any sort of URI resolution outside that method seems to defeat the purpose of that method then?

In my opinion every provider should know how to pick things out from their ConnectionDetails instead of a centralized struct that has a switch statement in it?

Reply:
It appears that for every supported external storage, we now have to provide a wrapper to be able to cast it to "sink"?
external-gcs, externa-azure, external-webhook... 
It just seems to be pretty verbose and for no benefit I can see.

Existing method (getSink) already knows how to parse every kind of (changefeed) URI.  It already handles
all sort of errors -- checks to make sure scheme is supported; etc, etc.
Why can't "external sink" just resolve whatever is in the connection detail and just call getSink on that?
Why do we need anything more than that?

Reply:
> It appears that for every supported external storage, we now have to provide a wrapper

Yes, this was the model that external storage/kms set out with. I agree about the verbosity of it, but then I think we need to rethink the proto-model in `connection.proto`. A connection detail per provider becomes useful when we will implement methods like `DisplayString` because each provider has a unique set of params they would like to redact. So being able to just do something like:

```
details := externalconn.MakeDetails(connectionpb.Details)
return details.DisplayString()
```

when implementing `SHOW` seems attractive. But ignoring status quo for a minute am I understanding correctly that what you're suggesting is roughly:

1. A single proto in `connection.proto` called `SinkDetails` or something. This will have the URI, and all other sink-specific configurations. So this will be a shared proto by **all** cdc sinks.

2. Every sink provider will still have to register a `scheme` and a parse+validate method that goes from URI -> ConnectionDetails for `CREATE EXTERNAL CONNECTION` to work. The difference from the current implementation is that the ConnectionDetails these methods create will all be of type `SinkDetails`. 

3. We will have an "external sink" that will implement the methods of the `ConnectionDetails` interface. So `Dial`, `ConnectionType`, `DisplayString` where some of these methods will have switch statements that based on the concrete Provider (kafka, pubsub etc.) will call custom logic.

Now today these constructors that implement `interface ConnectionDetails` are registered based on concrete provider type (nodelocal, kafka, pubusb etc.), but this will no longer be possible since we have a single "external sink" for all providers. So we need to have a way to say oh these opaque proto bytes need to go to this "external sink" constructor, vs these bytes go to the "external storage" constructor. So maybe we need to change the proto to have an additional ConnectionImpl which is what is used to register the constructors?


```
enum ConnectionImpl {
   ExternalStorage
   KMS
   Sink
}

message ConnectionDetails {
  ConnectionImpl impl = 1;
  ConnectionProvider provider = 2;

  oneof details {
    ExternalStorageDetails es = 3;
    KMSDetails kms = 4;
    SinkDetails sink = 5;
  }
}

enum ConnectionProvider {
  Unknown = 0;

  // ExternalStorage providers.
  nodelocal = 1;

  // KMS providers.
  gs_kms = 2;

  // Sink providers.
  kafka = 3;
}

message SinkDetails {
   string URI = 1;
   string json_config = 2;
}

message ExternalStorageDetails {
   string URI = 1;
}

message KMSDetails {
   string URI = ;
}
```

Reply:
Spoke offline; look forward to seeing the next version.